http-gateway/pkg/server/topic_existence_checker.go

```diff
@@ -33,29 +43,48 @@ func NewAlwaysTrueTopicExistenceChecker() TopicExistenceChecker {
 // NewRiffTopicExistenceChecker configures a TopicExistenceChecker using the
 // provided Clientset.
 func NewRiffTopicExistenceChecker(clientSet *versioned.Clientset) TopicExistenceChecker {
-	return &riffTopicExistenceChecker{client: clientSet}
-}
+	riffInformerFactory := informers.NewSharedInformerFactory(clientSet, time.Second*30)
+	topicInformer := riffInformerFactory.Projectriff().V1alpha1().Topics()
 
-func (tec *alwaysTrueTopicExistenceChecker) TopicExists(namespace string, topicName string) (bool, error) {
-	return true, nil
-}
+	knownTopics := make(map[string]ignoredValue)
 
-// TopicExists checks to see if Kubernetes is aware of a Riff Topic in a namespace.
-// If the topic exists, it returns (true, nil)
-// If the topic does not exist, it returns (false, nil)
-// If there is an unexpected error, it returns (false, err), where 'err' is the unexpected
-// error that was encountered.
-func (tec *riffTopicExistenceChecker) TopicExists(namespace string, topicName string) (bool, error) {
+	topicInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err != nil {
+				log.Printf("AddFunc had an error for key '%s': %+v", err)
+			}
 
-	_, err := tec.client.ProjectriffV1alpha1().Topics(namespace).Get(topicName, v1.GetOptions{})
+			knownTopics[key] = ignoredValue{}
+			log.Printf("Added topic to internal map: %+v", key)
+		},
+
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err != nil {
+				log.Printf("DeleteFunc had an error for key '%s': %+v", err)
+			}
+
+			delete(knownTopics, key)
+			log.Printf("Removed topic from internal map: %+v", key)
+		},
+	})
+
+	done := make(chan struct{}) //TODO: ideally, this would be the same channel used by the gateway itself
+	go topicInformer.Informer().Run(done)
+
+	return &riffTopicExistenceChecker{topicInformer: topicInformer, knownTopics: knownTopics}
+}
+
+func (tec *alwaysTrueTopicExistenceChecker) TopicExists(namespace string, topicName string) bool {
+	return true
+}
 
-	if err != nil {
-		if k8serrors.IsNotFound(err) {
-			return false, nil
-		}
+// TopicExists checks to see if Kubernetes is aware of a riff Topic in a namespace.
+func (tec *riffTopicExistenceChecker) TopicExists(namespace string, topicName string) bool {
+	topicKey := fmt.Sprintf("%s/%s", namespace, topicName)
 
-		return false, err
-	}
+	_, exists := tec.knownTopics[topicKey]
```

Comment:
Not thread safe. Use a mutex.

Reply:
Updated. I also kicked around using `RWLock`, but it was measurably slower (approx 20%). Also looked at `sync.Map`, but heeded the documentation's guidance not to use it unless strictly necessary.

Reply:
Did you try `RWMutex`? I don't think the benefits, if any, will be significant, but this would reduce contention between goroutines handling concurrent HTTP requests

Reply:
My mistake -- I wrote `RWLock` but meant `RWMutex`. It showed up as _slower_, which was a counterintuitive outcome.

Doing an eyeball comparison of `Mutex` and `RWMutex` code shows what appears to be an important difference. `Mutex.Lock()` has [a fast-path that uses a compare-and-swap operation](https://golang.org/src/sync/mutex.go#L73) to acquire the lock, whereas `RWMutex.RLock()` does not have such a fast path ([it atomically increments an int](https://golang.org/src/sync/rwmutex.go#L48)). My hunch is that the `Mutex.Lock()` version is compiling down to take advantage of CPU compare-and-swap instructions, which would be very fast in a lightly-contended lock. Why the `RWMutex` isn't being done with a CAS is a bit beyond my level of understanding.

And having thought further, the lock here _should_ be lightly-contended. The write operations should happen as fast as topics are added or removed, whereas the read operations will be per-message. So it seems as though the fast-path implementation of plain `Mutex` dominates the lock-contention advantage of `RWMutex`.